File: projects/composablekernel/python/ck_dsl_c/tests/parity/elementwise_emit.c
/* Copyright (c) Advanced Micro Devices, Inc., or its affiliates.
 * SPDX-License-Identifier: MIT
 *
 * tests/parity/elementwise_emit.c -- C-side emitter for the elementwise parity
 * harness. Selects one of N sampled ElementwiseSpec configs by argv[1] (the
 * config index), builds ckc_elementwise_spec_t identically to the Python
 * emitter elementwise_emit.py, lowers via ckc_elementwise_lower_to_llvm
 * (arch gfx950, flavor AUTO) and prints the .ll to stdout so the two outputs
 * can be byte-compared.
 */
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "ckc/ir.h"
#include "ckc/ir_serialize.h"
#include "ckc/lower_llvm.h"
#include "ckc/verify.h"
#include "ckc/instance_elementwise.h"

/* Fill `spec` for config index `idx`. Returns 0 on success, -1 if unknown. */
static int make_spec(int idx, ckc_elementwise_spec_t *spec) {
    *spec = ckc_elementwise_spec_default();

    switch (idx) {
    case 0:
        spec->op = "relu";  spec->dtype = "f16";  spec->block_size = 256; spec->vec = 8;
        break;
    case 1:
        spec->op = "relu";  spec->dtype = "bf16"; spec->block_size = 256; spec->vec = 8;
        break;
    case 2:
        spec->op = "add";   spec->dtype = "f16";  spec->block_size = 128; spec->vec = 4;
        break;
    case 3:
        spec->op = "add";   spec->dtype = "f16";  spec->block_size = 512; spec->vec = 2;
        break;
    case 4:
        spec->op = "silu";  spec->dtype = "bf16"; spec->block_size = 64;  spec->vec = 8;
        break;
    case 5:
        spec->op = "gelu_tanh"; spec->dtype = "f16"; spec->block_size = 1024; spec->vec = 4;
        break;
    default:
        return -1;
    }
    return 0;
}

int main(int argc, char **argv) {
    if (argc < 2) {
        fprintf(stderr, "usage: %s <config_index>\n", argv[0]);
        return 2;
    }
    int idx = atoi(argv[1]);
    const char *mode = (argc > 2) ? argv[2] : "ll";

    ckc_elementwise_spec_t spec;
    if (make_spec(idx, &spec) != 0) {
        fprintf(stderr, "unknown config index %d\n", idx);
        return 2;
    }

    if (strcmp(mode, "ll") == 0) {
        char *llvm_text = NULL;
        char err[CKC_ERR_MSG_CAP];
        err[0] = 0;
        ckc_status_t st = ckc_elementwise_lower_to_llvm(
            &spec, "gfx950", CKC_LLVM_FLAVOR_AUTO, &llvm_text, err, sizeof err);
        if (st != CKC_OK || !llvm_text) {
            fprintf(stderr, "lower failed: status=%d err=%s\n", (int)st, err);
            return 1;
        }
        fputs(llvm_text, stdout);
        free(llvm_text);
    } else if (strcmp(mode, "ir") == 0) {
        ckc_ir_builder_t b;
        ckc_kernel_def_t *kernel = ckc_build_elementwise_new(&b, &spec);
        if (!kernel) {
            fprintf(stderr, "build failed: %s\n", ckc_ir_builder_error(&b));
            ckc_ir_builder_free(&b);
            return 1;
        }
        char *t = NULL;
        ckc_status_t st = ckc_ir_serialize(kernel, &t);
        if (st != CKC_OK || !t) {
            fprintf(stderr, "ir_serialize failed: status=%d\n", (int)st);
            ckc_ir_builder_free(&b);
            return 1;
        }
        fputs(t, stdout);
        free(t);
        ckc_ir_builder_free(&b);
    } else if (strcmp(mode, "verify") == 0) {
        ckc_ir_builder_t b;
        ckc_kernel_def_t *kernel = ckc_build_elementwise_new(&b, &spec);
        if (!kernel) {
            fprintf(stderr, "build failed: %s\n", ckc_ir_builder_error(&b));
            ckc_ir_builder_free(&b);
            return 1;
        }
        ckc_diag_t *d = NULL;
        size_t n = 0;
        ckc_verify(kernel, &d, &n);
        for (size_t i = 0; i < n; i++) {
            char *s = ckc_diag_to_string(&d[i]);
            if (s) { puts(s); free(s); }
        }
        ckc_diags_free(d, n);
        ckc_ir_builder_free(&b);
    } else {
        fprintf(stderr, "unknown mode %s\n", mode);
        return 2;
    }
    return 0;
}

File: projects/composablekernel/python/ck_dsl_c/tests/parity/elementwise_emit.py
#!/usr/bin/env python3
# Copyright (c) Advanced Micro Devices, Inc., or its affiliates.
# SPDX-License-Identifier: MIT
#
# tests/parity/elementwise_emit.py -- Python reference emitter for the
# elementwise parity harness. Selects one of N sampled ElementwiseSpec configs
# by argv[1] (the config index), builds the kernel via build_elementwise and
# prints lower_kernel_to_llvm(kernel, arch='gfx950') to stdout so it can be
# byte-compared with the C emitter elementwise_emit.c.
import sys

from ck_dsl.instances.common.elementwise import (
    ElementwiseSpec,
    build_elementwise,
)
from ck_dsl import lower_kernel_to_llvm
from ck_dsl.core.ir_serialize import serialize
from ck_dsl.core.verify import verify


def _spec(idx: int) -> ElementwiseSpec:
    if idx == 0:
        return ElementwiseSpec(op="relu", dtype="f16", block_size=256, vec=8)
    if idx == 1:
        return ElementwiseSpec(op="relu", dtype="bf16", block_size=256, vec=8)
    if idx == 2:
        return ElementwiseSpec(op="add", dtype="f16", block_size=128, vec=4)
    if idx == 3:
        return ElementwiseSpec(op="add", dtype="f16", block_size=512, vec=2)
    if idx == 4:
        return ElementwiseSpec(op="silu", dtype="bf16", block_size=64, vec=8)
    if idx == 5:
        return ElementwiseSpec(op="gelu_tanh", dtype="f16", block_size=1024, vec=4)
    raise SystemExit(f"unknown config index {idx}")


def main() -> int:
    if len(sys.argv) < 2:
        sys.stderr.write("usage: elementwise_emit.py <config_index>\n")
        return 2
    idx = int(sys.argv[1])
    mode = sys.argv[2] if len(sys.argv) > 2 else "ll"
    spec = _spec(idx)
    kernel = build_elementwise(spec)
    if mode == "ll":
        text = lower_kernel_to_llvm(kernel, arch="gfx950")
        sys.stdout.write(text)
    elif mode == "ir":
        sys.stdout.write(serialize(kernel))
    elif mode == "verify":
        sys.stdout.write("".join(str(d) + "\n" for d in verify(kernel)))
    else:
        sys.stderr.write(f"unknown mode {mode}\n")
        return 2
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
